src/operator/window.rs

```diff
@@ -0,0 +1,424 @@
+//! Operators to organize time series data into windows.
+
+use crate::{
+    algebra::{IndexedZSet, NegByRef},
+    circuit::{
+        operator_traits::{Operator, TernaryOperator},
+        Circuit, OwnershipPreference, Scope, Stream,
+    },
+    trace::{cursor::Cursor, ord::OrdZSet, spine_fueled::Spine, Batch, BatchReader},
+};
+use deepsize::DeepSizeOf;
+use std::{borrow::Cow, cmp::max, marker::PhantomData, rc::Rc};
+
+impl<P, B> Stream<Circuit<P>, B>
+where
+    P: Clone + 'static,
+    B: IndexedZSet + DeepSizeOf,
+    B::Key: Ord + Clone,
+    B::Val: Ord + Clone,
+    B::R: NegByRef,
+{
+    // TODO: An even more general window method would read window bounds from a
+    // separate stream. This way, bounds could be derived from data.  The only
+    // complication is that we'll have to support ternary operators for this.
+
+    /// Extract a subset of values that fall within a moving window from a
+    /// stream of time-indexed values.
+    ///
+    /// This is a general form of the windowing operator that supports tumbling,
+    /// rolling windows, watermarks, etc., by relying on a user-supplied
+    /// function to compute window bounds at each clock cycle.
+    ///
+    /// This operator maintains the window **incrementally**, i.e., it outputs
+    /// changes to the contents of the window at each clock cycle.  The
+    /// complete contents of the window can be computed by integrating the
+    /// output stream.
+    ///
+    /// # Arguments
+    ///
+    /// * `self` - stream of indexed Z-sets (indexed by time).  The notion of
+    ///   time here is distinct from the DBSP logical time and can be modeled
+    ///   using any type that implements `Ord`.
+    ///
+    /// * `bounds` - stream that contains window bounds to use at each clock
```

Comment:
why don't you draw the circuit generated by this function?
The inner operator is ternary, but this generates a binary circuit, with an inner feedback loop.